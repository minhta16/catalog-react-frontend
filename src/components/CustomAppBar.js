import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  AppBar, Toolbar, IconButton, Typography,
} from '@material-ui/core';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import HideOnScroll from './HideOnScroll';
import LoginButton from './LoginButton';

class CustomAppBar extends Component {
  state = {};

  render() {
    const { color, currentUser } = this.props;
    return (
      <HideOnScroll>
        <AppBar color={color}>
          <Toolbar>
            <IconButton edge="start" color="inherit" component={Link} exact="true" to="/">
              <img src="/img/logo.svg" alt="site logo" />
            </IconButton>
            <Typography style={{ flexGrow: '1' }} />
            <LoginButton currentUser={currentUser} />
          </Toolbar>
        </AppBar>
      </HideOnScroll>
    );
  }
}

CustomAppBar.propTypes = {
  color: PropTypes.string.isRequired,
  currentUser: PropTypes.shape({
    username: PropTypes.string,
    password: PropTypes.string,
    token: PropTypes.string,
  }),
};

CustomAppBar.defaultProps = {
  currentUser: { username: '', password: '', token: '' },
};

const mapStateToProps = state => ({
  currentUser: state.currentUser,
});

export default connect(mapStateToProps)(CustomAppBar);
