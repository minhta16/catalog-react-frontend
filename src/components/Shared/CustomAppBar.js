import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { AppBar, Toolbar, IconButton, Typography } from '@material-ui/core';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import {
  signIn as signInRedux,
  signOut as signOutRedux,
  clearError as clearErrorRedux,
} from 'actions/users';
import { selectCurrentUser, selectLoginErrorMessage } from 'selectors/users';
import LoginButton from './LoginButton';
import LoginDialog from './LoginDialog';

export class CustomAppBar extends Component {
  state = {
    open: false,
  };

  handleOpenDialog = () => {
    this.setState({
      open: true,
    });
  };

  handleCloseDialog = () => {
    const { clearError } = this.props;
    clearError();
    this.setState({
      open: false,
    });
  };

  /**
   * close the login dialog if currentUser token exists, meaning that the user just logged in
   */
  componentDidUpdate = (prevProps) => {
    const { currentUser } = this.props;
    if (!prevProps.currentUser.token && currentUser.token) {
      this.handleCloseDialog();
    }
  };

  render() {
    const { open } = this.state;
    const { color, currentUser, signIn, signOut, errorMessage } = this.props;
    return (
      <AppBar color={color}>
        <Toolbar>
          <IconButton edge="start" color="inherit" component={Link} exact="true" to="/">
            <img src="/img/logo.svg" alt="site logo" />
          </IconButton>
          <Typography style={{ flexGrow: '1' }} />
          <LoginButton
            id="custom-app-bar-login-button"
            variant="contained"
            color="primary"
            currentUser={currentUser}
            onClick={this.handleOpenDialog}
            signOut={signOut}
          />
          <LoginDialog
            open={open}
            onClose={this.handleCloseDialog}
            onClick={signIn}
            errorMessage={errorMessage}
          />
        </Toolbar>
      </AppBar>
    );
  }
}

CustomAppBar.propTypes = {
  color: PropTypes.string.isRequired,
  currentUser: PropTypes.object,
  signIn: PropTypes.func.isRequired,
  signOut: PropTypes.func.isRequired,
  errorMessage: PropTypes.array,
  clearError: PropTypes.func.isRequired,
};

CustomAppBar.defaultProps = {
  currentUser: {},
  errorMessage: [],
};

export const mapSelectorToProps = (state) => ({
  currentUser: selectCurrentUser(state),
  errorMessage: selectLoginErrorMessage(state),
});

export const mapDispatchToProps = {
  signIn: signInRedux,
  signOut: signOutRedux,
  clearError: clearErrorRedux,
};

export default connect(
  mapSelectorToProps,
  mapDispatchToProps,
)(CustomAppBar);
