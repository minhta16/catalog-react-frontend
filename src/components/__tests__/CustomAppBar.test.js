/* eslint-disable no-undef */
import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import CustomAppBar from 'components/CustomAppBar';

const mockStore = configureMockStore();
const store = mockStore({});

configure({ adapter: new Adapter() });

describe('components/CustomAppBar', () => {
  let wrapper;
  let props;
  // let warning;

  // const update = () => {
  //   wrapper.update();
  //   warning = wrapper.find(Warning);
  // };

  const setup = () => {
    wrapper = shallow(
      <Provider store={store}>
        <CustomAppBar {...props} />
      </Provider>,
    );
  };

  beforeEach(() => {
    props = {
      color: 'primary',
      currentUser: {
        username: 'minh',
        password: 'minh',
        token: 'abc',
      },
    };
  });

  it('should render correctly', () => {
    setup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
