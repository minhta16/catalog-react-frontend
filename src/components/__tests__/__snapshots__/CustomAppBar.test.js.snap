// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CustomAppBar should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <Connect(CustomAppBar)
    color="primary"
    currentUser={
      Object {
        "password": "minh",
        "token": "abc",
        "username": "minh",
      }
    }
  />
</ContextProvider>
`;
